sable_ircd/src/command/handlers/chathistory.rs

```diff
@@ -197,20 +198,48 @@ async fn list_targets<'a>(
 }
 
 fn send_history_entries<'a>(
-    server: &ClientServer,
-    into: impl MessageSink,
+    _server: &ClientServer,
+    conn: impl MessageSink,
     target: &str,
-    entries: impl IntoIterator<Item = HistoryLogEntry>,
+    entries: impl IntoIterator<Item = HistoricalEvent>,
 ) -> CommandResult {
-    let batch = into
+    let batch = conn
         .batch("chathistory", ClientCapability::Batch)
         .with_arguments(&[target])
         .start();
 
     for entry in entries {
-        // Ignore errors here; it's possible that a message has been expired out of network state
-        // but a reference to it still exists in the history log
-        let _ = server.send_item(&entry, &batch, &entry);
+        match entry {
+            HistoricalEvent::Message {
+                id,
+                source,
+                source_account,
+                target: _, // assume it's the same as the one we got as parameter
+                message_type,
+                text,
+            } => {
+                let (timestamp, _) = id
+                    .get_timestamp()
+                    .expect("message has non-v7 UUID")
+                    .to_unix();
+                let msg = message::Message::new(&source, &target, message_type, &text)
+                    .with_tag(server_time::server_time_tag(
+                        i64::try_from(timestamp).unwrap_or(i64::MAX),
+                    ))
```

Comment:
This is unsound. I stumbled upon this issue after running irctest a dozen times:

1. client sends a message, gets this echo: `@time=2025-10-27T17:18:50.000Z;msgid=019a26ae-048f-7352-80cb-82d9d3ded30f :bar!username@localhost PRIVMSG #chanac0a9c4d540ce04f7a35da65 :this is message 3` (note that `019a26ae-048f-7352-80cb-82d9d3ded30f`'s timestamp is `2025-10-27T17:18:49.999Z`)
2. sable_history writes this: `Message { id: 019a26ae-048f-7352-80cb-82d9d3ded30f, source_user: 1, target_channel: 770697994130427908, text: "this is message 3", message_type: Privmsg }`
3. client asks for chathistory
4. ircd asks sable_history, gets `Message { id: MessageId(Uuid7(019a26ae-048f-7352-80cb-82d9d3ded30f)), source: "bar!username@localhost", source_account: None, target: "#chanac0a9c4d540ce04f7a35da65", message_type: Privmsg, text: "this is message 3" }` back
5. ircd replies this to the client: `@time=2025-10-27T17:18:49.000Z;msgid=019a26ae-048f-7352-80cb-82d9d3ded30f;batch=e5286d57605d2baa2395dc78c4d7d042 :bar!username@localhost PRIVMSG #chanac0a9c4d540ce04f7a35da65 :this is message 3`

Reply:
> IMO we should build the UUID in a way that ensures it matches the timestamp.

except we can't because the UUID is part of the message, which is built before we have a timestamp from the NewMessage state change

Reply:
Fixed by storing the update timestamp as its own column in the DB